In [1]:
import numpy as np
import pandas as pd
from sklearn import linear_model
import matplotlib.pyplot as plt
from sqlalchemy import create_engine
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'weatherinszeged'

In [2]:
engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

weather = pd.read_sql_query('select * from weatherinszeged',con=engine)

engine.dispose()

weather.head(10)

date        summary preciptype  temperature  \
0 2006-03-31 22:00:00+00:00  Partly Cloudy       rain     9.472222   
1 2006-03-31 23:00:00+00:00  Partly Cloudy       rain     9.355556   
2 2006-04-01 00:00:00+00:00  Mostly Cloudy       rain     9.377778   
3 2006-04-01 01:00:00+00:00  Partly Cloudy       rain     8.288889   
4 2006-04-01 02:00:00+00:00  Mostly Cloudy       rain     8.755556   
5 2006-04-01 03:00:00+00:00  Partly Cloudy       rain     9.222222   
6 2006-04-01 04:00:00+00:00  Partly Cloudy       rain     7.733333   
7 2006-04-01 05:00:00+00:00  Partly Cloudy       rain     8.772222   
8 2006-04-01 06:00:00+00:00  Partly Cloudy       rain    10.822222   
9 2006-04-01 07:00:00+00:00  Partly Cloudy       rain    13.772222   

   apparenttemperature  humidity  windspeed  windbearing  visibility  \
0             7.388889      0.89    14.1197        251.0     15.8263   
1             7.227778      0.86    14.2646        259.0     15.8263   
2             9.377778      0.89     3.9284        204.0     14.9569   
3             5.944444      0.83    14.1036        269.0     15.8263   
4             6.977778      0.83    11.0446        259.0     15.8263   
5             7.111111      0.85    13.9587        258.0     14.9569   
6             5.522222      0.95    12.3648        259.0      9.9820   
7             6.527778      0.89    14.1519        260.0      9.9820   
8            10.822222      0.82    11.3183        259.0      9.9820   
9            13.772222      0.72    12.5258        279.0      9.9820   

   loudcover  pressure                       dailysummary  
0        0.0   1015.13  Partly cloudy throughout the day.  
1        0.0   1015.63  Partly cloudy throughout the day.  
2        0.0   1015.94  Partly cloudy throughout the day.  
3        0.0   1016.41  Partly cloudy throughout the day.  
4        0.0   1016.51  Partly cloudy throughout the day.  
5        0.0   1016.66  Partly cloudy throughout the day.  
6        0.0   1016.72  Partly cloudy throughout the day.  
7        0.0   1016.84  Partly cloudy throughout the day.  
8        0.0   1017.37  Partly cloudy throughout the day.  
9        0.0   1017.22  Partly cloudy throughout the day.

Build a linear regression model where your target variable is the difference between the apparenttemperature and the temperature. As explanatory variables, use humidity and windspeed. Now, estimate your model using OLS. Are the estimated coefficients statistically significant? Are the signs of the estimated coefficients in line with your previous expectations? 
Interpret the estimated coefficients. What are the relations between the target and the explanatory variables?

In [5]:
# target variable = apparenttemp. - temp.
Y = weather['apparenttemperature'] - weather['temperature']

# features = humidity and windspeed
X = weather[['humidity', 'windspeed']]

lrm = linear_model.LinearRegression()

lrm.fit(X, Y)

print('\nCoefficients: \n', lrm.coef_)
print('\nIntercept: \n', lrm.intercept_)


Coefficients: 
 [-3.02918594 -0.11929075]

Intercept: 
 2.4381054151876933


In [6]:
import statsmodels.api as sm

# We need to manually add a constant
X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Thu, 17 Oct 2019   Prob (F-statistic):               0.00
Time:                        08:49:09   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.4381      0.021    115.948      0.000       2.397       2.479
humidity      -3.0292      0.024   -126.479      0.000      -3.076      -2.982
windspeed     -0.1193      0.001   -176.164      0.000      -0.121      -0.118
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.267
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                          -0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Looking at the p-values for both of our features they both appear to be statistically significant as they are below 0.05. 
Compared to our previous model where we looked at just temperature as a target variable and included windbearing and pressure in the explanatory variables, both humidity and windspeed have decreased significantly in magnitude. They still have the same signs however. 
Interpreting the coefficients we see that for every 1 point change in humidity there is a 3.03 decrease in the target. For every 1 point change in windspeed there is a 0.12 decrease in the target. 

Next, include the interaction of humidity and windspeed to the model above and estimate the model using OLS. Are the coefficients statistically significant? Did the signs of the estimated coefficients for humidity and windspeed change? Interpret the estimated coefficients.

In [7]:
# target variable = apparenttemp. - temp.
Y = weather['apparenttemperature'] - weather['temperature']


# interaction of humidity and windspeed
weather['humidity_windspeed'] = weather.humidity * weather.windspeed

# features = humidity, windspeed and interaction of both 
X = weather[['humidity', 'windspeed', 'humidity_windspeed']]

lrm = linear_model.LinearRegression()

lrm.fit(X, Y)

print('\nCoefficients: \n', lrm.coef_)
print('\nIntercept: \n', lrm.intercept_)

import statsmodels.api as sm

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()


Coefficients: 
 [ 0.17751219  0.09048213 -0.29711946]

Intercept: 
 0.08393631009782698


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Thu, 17 Oct 2019   Prob (F-statistic):               0.00
Time:                        08:59:47   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                  0.0839      0.033      2.511      0.012       0.018       0.149
humidity               0.1775      0.043      4.133      0.000       0.093       0.262
windspeed              0.0905      0.002     36.797      0.000       0.086       0.095
humidity_windspeed    -0.2971      0.003    -88.470      0.000      -0.304      -0.291
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.265
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                          -0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

All of the features are still statistically significant, however the signs of both humidity and windspeed changed. Now for every 1 point change in humidity or windspeed, we observe a 0.18 and 0.09 increase in the target respectively. Conversely, for every 1 point change in the interaction of the two, we see a 0.2971 decrease in the target variable. 